<a href="https://colab.research.google.com/github/AlexBB999/NLP/blob/master/31_6_Assignment_NLP_features_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
#pip install gensim

In [2]:
import numpy as np
import pandas as pd
import sklearn
import spacy
import re
import nltk
from nltk.corpus import gutenberg
import gensim
import warnings
warnings.filterwarnings("ignore")

nltk.download('gutenberg')
!python -m spacy download en

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Unzipping corpora/gutenberg.zip.
✔ Download and installation successful
You can now load the model via spacy.load('en_core_web_sm')
✔ Linking successful
/usr/local/lib/python3.6/dist-packages/en_core_web_sm -->
/usr/local/lib/python3.6/dist-packages/spacy/data/en
You can now load the model via spacy.load('en')


**UTITLITY FUNCTION FOR STANDARD TEXT CLEANING**

In [0]:
# utility function for standard text cleaning
def text_cleaner(text):
    # visual inspection identifies a form of punctuation spaCy does not
    # recognize: the double dash '--'.  Better get rid of it now!
    text = re.sub(r'--',' ',text)
    text = re.sub("[\[].*?[\]]", "", text)
    text = re.sub(r"(\b|\s+\-?|^\-?)(\d+|\d*\.\d+)\b", " ", text)
    text = ' '.join(text.split())
    return text

**LOAD AND CLEAN THE DATA**

In [0]:
# load and clean the data
persuasion = gutenberg.raw('austen-persuasion.txt')
alice = gutenberg.raw('carroll-alice.txt')

# the chapter indicator is idiosyncratic
persuasion = re.sub(r'Chapter \d+', '', persuasion)
alice = re.sub(r'CHAPTER .*', '', alice)
    
alice = text_cleaner(alice)
persuasion = text_cleaner(persuasion)

**PARSE THE CLEANED DATA**

In [0]:
# parse the cleaned novels. This can take a bit.
nlp = spacy.load('en')
alice_doc = nlp(alice)
persuasion_doc = nlp(persuasion)

**GROUP INTO SENTENCES**

In [6]:
# group into sentences
alice_sents = [[sent, "Carroll"] for sent in alice_doc.sents]
persuasion_sents = [[sent, "Austen"] for sent in persuasion_doc.sents]

# combine the sentences from the two novels into one data frame
sentences = pd.DataFrame(alice_sents + persuasion_sents, columns = ["text", "author"])
sentences.head()

text   author
0  (Alice, was, beginning, to, get, very, tired, ...  Carroll
1  (So, she, was, considering, in, her, own, mind...  Carroll
2  (There, was, nothing, so, VERY, remarkable, in...  Carroll
3                                      (Oh, dear, !)  Carroll
4                         (I, shall, be, late, !, ')  Carroll

**GET RID OF STOP WORDS AND PUNCTUATION**

In [0]:
# get rid off stop words and punctuation
# and lemmatize the tokens
for i, sentence in enumerate(sentences["text"]):
    sentences.loc[i, "text"] = [token.lemma_ for token in sentence if not token.is_punct and not token.is_stop]

**NOW READY TO VECTORIZE USING WORD2VEC**



**The Word2Vec class has several parameters**.

We set the following parameters:

**workers**=4: We set the number of threads to run in parallel to 4 (make sense if your computer has available computing units).

**min_count**=1: We set the minimum word count threshold to 1.

**window**=6: We set the number of words around target word to consider to 6.

**sg**=0: We use CBOW because our corpus is small.

**sample**=1e-3: We penalize frequent words.

**size**=100: We set the word vector length to 100.

**hs**=1: We use hierarchical softmax.

In [0]:
# train word2vec on the the sentences
model = gensim.models.Word2Vec(
    sentences["text"],
    workers=4,
    min_count=1,
    window=6,
    sg=0,
    sample=1e-3,
    size=100,
    hs=1
)

Before jumping in the machine learning model for prediction,

 **let's play with our word2vec word representation we just trained**. 
 
 **specifically, we'll look into**:

The first five words that are closer to lady.

The word that doesn't fit in list: dad dinner mom aunt uncle.

The similarity score of woman and man.

The similarity score of horse and cat.

Note that all of the above calculations are based on the word2vec representations of the words we just trained above.

In [9]:
print(model.most_similar(positive=['lady', 'man'], negative=['woman'], topn=5))
print(model.doesnt_match("dad dinner mom aunt uncle".split()))
print(model.similarity('woman', 'man'))
print(model.similarity('horse', 'cat'))

[('heart', 0.9986150860786438), ('head', 0.9980312585830688), ('receive', 0.9979260563850403), ('send', 0.9978547096252441), ('compare', 0.9978461265563965)]
dinner
0.99793494
0.91907066


Well, the results make sense to some degree but it's obvious that our representations aren't perfect.

 **This is because our corpus is small**.
 
In order to get more meaningful results, we need to train word2vec representations using much larger corpuses.


Now, **let's create our numerical features using the word2vec representations of the words**. 

**In the following, we get the word2vec vectors of each word in a sentence and take the average of all the vectors in the high dimensional space** (in our case it's 100).

**So, as a result, we'll have a vector of 100 dimensions as the feature for a sentence**.
 
 **We then use each dimension as a separate feature**
 
  **which means that in our final dataset we'll have 100 numerical feature**

In [10]:
word2vec_arr = np.zeros((sentences.shape[0],100))

for i, sentence in enumerate(sentences["text"]):
    word2vec_arr[i,:] = np.mean([model[lemma] for lemma in sentence], axis=0)

word2vec_arr = pd.DataFrame(word2vec_arr)
sentences = pd.concat([sentences[["author", "text"]],word2vec_arr], axis=1)
sentences.dropna(inplace=True)

sentences.head()

author  ...        99
0  Carroll  ... -0.131617
1  Carroll  ... -0.098131
2  Carroll  ... -0.150146
3  Carroll  ... -0.113956
4  Carroll  ... -0.101913

[5 rows x 102 columns]

**This is a dataset format that we like**.

 **Now, we're ready to jump into modeling step with our features**. 

##**Word2vec in action**

**Notice that we now have a dataset where the columns named from 0 to 99 are the features we'll use in the following models**. 

We use the same models that we built in the previous checkpoints to predict the author of a sentence:

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.model_selection import train_test_split

Y = sentences['author']
X = np.array(sentences.drop(['text','author'], 1))

# We split the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.4, random_state=123)

# Models
lr = LogisticRegression()
rfc = RandomForestClassifier()
gbc = GradientBoostingClassifier()

lr.fit(X_train, y_train)
rfc.fit(X_train, y_train)
gbc.fit(X_train, y_train)

print("----------------------Logistic Regression Scores----------------------")
print('Training set score:', lr.score(X_train, y_train))
print('\nTest set score:', lr.score(X_test, y_test))

print("----------------------Random Forest Scores----------------------")
print('Training set score:', rfc.score(X_train, y_train))
print('\nTest set score:', rfc.score(X_test, y_test))

print("----------------------Gradient Boosting Scores----------------------")
print('Training set score:', gbc.score(X_train, y_train))
print('\nTest set score:', gbc.score(X_test, y_test))

----------------------Logistic Regression Scores----------------------
Training set score: 0.7485540334855403

Test set score: 0.7593607305936073
----------------------Random Forest Scores----------------------
Training set score: 0.9914764079147641

Test set score: 0.7986301369863014
----------------------Gradient Boosting Scores----------------------
Training set score: 0.8867579908675799

Test set score: 0.8082191780821918


The scores aren't great compared to the scores of the previous checkpoints.

**The main reason is the small size of our corpus**.

**So, let's use word2vec vectors that are trained on a very large corpus**.

 For this, we use pre-trained vectors released by Google. Google released a quite large word2vec vectors that are trained on around 100 billion words from Google News dataset. 
 
 
**Their corpus contains 3 million words and the word vectors they trained have 300 features each**.

We'll download the **pre-trained vectors** from this address: https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz. Note that the download and the following codes take some time. So, we recommend running the following cells in Google Colab

In [0]:
# Load Google's pre-trained Word2Vec model.
model_pretrained = gensim.models.KeyedVectors.load_word2vec_format(
    'https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz', binary=True)

**Now, we have the pre-trained vectors in a variable called model_pretrained**. 

**Next, we look for the vector representations of the words in our corpus**.

 **For simplicity, if a word in a sentence can't be found in the vocabulary of these pre-trained vectors,**
 
 **we just simply drop those sentences from our dataset**.
 
 ou can follow alternative approaches if you like

In [0]:
word2vec_arr = np.zeros((sentences.shape[0],300))

for i, sentence in enumerate(sentences["text"]):
  try:
    word2vec_arr[i,:] = np.mean([model_pretrained[lemma] for lemma in sentence], axis=0)
  except KeyError:
    word2vec_arr[i,:] = np.full((1,300), np.nan) #fill in with nan -- this vector will be dropped on line #12
    continue

word2vec_arr = pd.DataFrame(word2vec_arr)
sentences = pd.concat([sentences[["author", "text"]],word2vec_arr], axis=1)
sentences.dropna(inplace=True)

print("Shape of the dataset: {}".format(sentences.shape))
sentences.head()

Shape of the dataset: (4610, 302)


author  ...       299
0  Carroll  ... -0.064371
1  Carroll  ...  0.000908
2  Carroll  ...  0.029114
3  Carroll  ...  0.186798
4  Carroll  ...  0.087646

[5 rows x 302 columns]

As a result, we have a dataset of 4114 rows and 300 features (excluding the text and the author columns).

Now, we can run our classifiers using this dataset

**SEEMS EACH RUN PRODUCES A SLIGHTLY DIFF NUMBER OF ROWS**

**OR ELSE CORPUS HAS BEEN UPDATED**

In [0]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.model_selection import train_test_split

Y = sentences['author']
X = np.array(sentences.drop(['text','author'], 1))

# We split the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.4, random_state=123)

# Models
lr = LogisticRegression()
rfc = RandomForestClassifier()
gbc = GradientBoostingClassifier()

lr.fit(X_train, y_train)
rfc.fit(X_train, y_train)
gbc.fit(X_train, y_train)

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=None, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [0]:
print("----------------------Logistic Regression Scores----------------------")
print('Training set score:', lr.score(X_train, y_train))
print('\nTest set score:', lr.score(X_test, y_test))

print("----------------------Random Forest Scores----------------------")
print('Training set score:', rfc.score(X_train, y_train))
print('\nTest set score:', rfc.score(X_test, y_test))

print("----------------------Gradient Boosting Scores----------------------")
print('Training set score:', gbc.score(X_train, y_train))
print('\nTest set score:', gbc.score(X_test, y_test))

----------------------Logistic Regression Scores----------------------
Training set score: 0.8897324656543746

Test set score: 0.8449023861171366
----------------------Random Forest Scores----------------------
Training set score: 0.9924078091106291

Test set score: 0.7977223427331888
----------------------Gradient Boosting Scores----------------------
Training set score: 0.9508315256688359

Test set score: 0.8156182212581344


**IMPROVED RESULTS**

#**ASSIGNMENTS**

**Train your own word2vec representations** as we did in our first example in the checkpoint.

But, you need to experiment with the hyperparameters of the vectorization step

**Modify the hyperparameters and run the classification models again**

**Can you wrangle any improvements**?

**MAYBE, ONCE  I UNDERSTAND THE PARAMETERS**

**HERE ARE ORIGINAL PARAMETERS:**



The Word2Vec class has several parameters.

We set the following parameters:

**workers**=4: We set the number of threads to run in parallel to 4 (make sense if your computer has available computing units).

**min_count**=1: We set the minimum word count threshold to 1.

**window**=6: We set the number of words around target word to consider to 6.

**sg**=0: We use CBOW because our corpus is small.

**sample**=1e-3: We penalize frequent words.

**size**=100: We set the word vector length to 100.

**hs**=1: We use hierarchical softmax.

I WILL CHANGE A FEW:

SG=1  SKIP-GRAM

HS=0

In [0]:
# train word2vec on the the sentences
model = gensim.models.Word2Vec(
    sentences["text"],
    workers=4,
    min_count=1,
    window=6,
    sg=1,
    sample=1e-3,
    size=100,
    hs=0
)

In [14]:
print(model.most_similar(positive=['lady', 'man'], negative=['woman'], topn=5))
print(model.doesnt_match("dad dinner mom aunt uncle".split()))
print(model.similarity('woman', 'man'))
print(model.similarity('horse', 'cat'))

[('brother', 0.9980732798576355), ('come', 0.99798583984375), ('sight', 0.9979526400566101), ('join', 0.997667133808136), ('Laconia', 0.9976547360420227)]
uncle
0.998033
0.9990655


**QUITE DIFFERENT**

In [15]:
word2vec_arr = np.zeros((sentences.shape[0],100))

for i, sentence in enumerate(sentences["text"]):
    word2vec_arr[i,:] = np.mean([model[lemma] for lemma in sentence], axis=0)

word2vec_arr = pd.DataFrame(word2vec_arr)
sentences = pd.concat([sentences[["author", "text"]],word2vec_arr], axis=1)
sentences.dropna(inplace=True)

sentences.head()

author  ...        99
0  Carroll  ... -0.095955
1  Carroll  ... -0.082511
2  Carroll  ... -0.095584
3  Carroll  ... -0.104728
4  Carroll  ... -0.094273

[5 rows x 102 columns]

**USE IN MODELS**

In [16]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.model_selection import train_test_split

Y = sentences['author']
X = np.array(sentences.drop(['text','author'], 1))

# We split the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.4, random_state=123)

# Models
lr = LogisticRegression()
rfc = RandomForestClassifier()
gbc = GradientBoostingClassifier()

lr.fit(X_train, y_train)
rfc.fit(X_train, y_train)
gbc.fit(X_train, y_train)

print("----------------------Logistic Regression Scores----------------------")
print('Training set score:', lr.score(X_train, y_train))
print('\nTest set score:', lr.score(X_test, y_test))

print("----------------------Random Forest Scores----------------------")
print('Training set score:', rfc.score(X_train, y_train))
print('\nTest set score:', rfc.score(X_test, y_test))

print("----------------------Gradient Boosting Scores----------------------")
print('Training set score:', gbc.score(X_train, y_train))
print('\nTest set score:', gbc.score(X_test, y_test))

----------------------Logistic Regression Scores----------------------
Training set score: 0.7367906066536204

Test set score: 0.7451076320939335
----------------------Random Forest Scores----------------------
Training set score: 0.9944553163731246

Test set score: 0.812133072407045
----------------------Gradient Boosting Scores----------------------
Training set score: 0.8956294846705806

Test set score: 0.8082191780821918


**OLD SCORES:**

----------------------Logistic Regression Scores----------------------
Training set score: 0.7485540334855403

Test set score: 0.7593607305936073

----------------------Random Forest Scores----------------------
Training set score: 0.9914764079147641

Test set score: 0.7986301369863014

----------------------Gradient Boosting Scores----------------------
Training set score: 0.8867579908675799

Test set score: 0.8082191780821918

**OVERALL ABOUT THE SAME**

LOGITIC REGRESSION TEST SCORE IMPROVED

RANDOM FOREST TEST WORSE

GRADIENT BOOST -- IDENTICAL TEST SCORE

**///////////////////////////////////////////////////////////////////**

**TRY ONCE MORE WITH A FEW CHANGES**

INCREASE WINDOW 

INCREASE SIZE

BACK TO HIERARCHICAL



In [0]:
# train word2vec on the the sentences
model = gensim.models.Word2Vec(
    sentences["text"],
    workers=4,
    min_count=1,
    window=7,
    sg=1,
    sample=1e-3,
    size=150,
    hs=1
)

In [18]:
print(model.most_similar(positive=['lady', 'man'], negative=['woman'], topn=5))
print(model.doesnt_match("dad dinner mom aunt uncle".split()))
print(model.similarity('woman', 'man'))
print(model.similarity('horse', 'cat'))

[('drawing', 0.941796600818634), ('state', 0.9320310950279236), ('gallant', 0.9194947481155396), ('remain', 0.9146274924278259), ('Harvilles', 0.9136101007461548)]
dinner
0.9044591
0.55610454


**CREATE DATASET WITH 100 NUMERICAL FEATURES**

In [20]:
word2vec_arr = np.zeros((sentences.shape[0],150))

for i, sentence in enumerate(sentences["text"]):
    word2vec_arr[i,:] = np.mean([model[lemma] for lemma in sentence], axis=0)

word2vec_arr = pd.DataFrame(word2vec_arr)
sentences = pd.concat([sentences[["author", "text"]],word2vec_arr], axis=1)
sentences.dropna(inplace=True)

sentences.head()

author  ...       149
0  Carroll  ... -0.171606
1  Carroll  ... -0.125782
2  Carroll  ... -0.130038
3  Carroll  ... -0.170898
4  Carroll  ... -0.150806

[5 rows x 152 columns]

In [21]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.model_selection import train_test_split

Y = sentences['author']
X = np.array(sentences.drop(['text','author'], 1))

# We split the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.4, random_state=123)

# Models
lr = LogisticRegression()
rfc = RandomForestClassifier()
gbc = GradientBoostingClassifier()

lr.fit(X_train, y_train)
rfc.fit(X_train, y_train)
gbc.fit(X_train, y_train)

print("----------------------Logistic Regression Scores----------------------")
print('Training set score:', lr.score(X_train, y_train))
print('\nTest set score:', lr.score(X_test, y_test))

print("----------------------Random Forest Scores----------------------")
print('Training set score:', rfc.score(X_train, y_train))
print('\nTest set score:', rfc.score(X_test, y_test))

print("----------------------Gradient Boosting Scores----------------------")
print('Training set score:', gbc.score(X_train, y_train))
print('\nTest set score:', gbc.score(X_test, y_test))

----------------------Logistic Regression Scores----------------------
Training set score: 0.8338590956887487

Test set score: 0.8170347003154574
----------------------Random Forest Scores----------------------
Training set score: 0.9922888187872415

Test set score: 0.8338590956887487
----------------------Gradient Boosting Scores----------------------
Training set score: 0.9120224325271644

Test set score: 0.8375394321766562


OLD SCORES:

----------------------Logistic Regression Scores---------------------- Training set score: 0.7485540334855403

Test set score: 0.7593607305936073

----------------------Random Forest Scores---------------------- Training set score: 0.9914764079147641

Test set score: 0.7986301369863014

----------------------Gradient Boosting Scores---------------------- Training set score: 0.8867579908675799

Test set score: 0.8082191780821918

**BETTER FOR LOGISTIC REGRESSION**

**LOWER FOR RF and GB**